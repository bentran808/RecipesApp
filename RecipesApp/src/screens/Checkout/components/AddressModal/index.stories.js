import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react-native';
import { address } from 'mocks';
import React from 'react';
import CenterView from 'storybook/stories/CenterView';
import AddressModal from '.';

storiesOf('Address Modal', module)
  .addDecorator((getStory) => <CenterView>{getStory()}</CenterView>)
  .add('default', () => (
    <AddressModal
      navigation={{
        navigate: action('navigate')
      }}
      data={address}
      modalVisible={true}
      onToggleModal={action('toggle-modal')}
      renderSeparator={() => null}
    />
  ));
