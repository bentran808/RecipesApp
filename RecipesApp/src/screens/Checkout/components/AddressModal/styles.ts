import { StyleSheet } from 'react-native';
import { width } from 'theme';

const styles = StyleSheet.create({
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0,0,0,0.5)'
  },
  modalView: {
    width: width - 50,
    backgroundColor: 'white',
    borderRadius: 20,
    paddingVertical: 20,
    alignItems: 'center',
    shadowColor: 'black',
    shadowOffset: {
      width: 0,
      height: 2
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5
  },
  addressWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingVertical: 10,
    width: width - 100
  }
});

export default styles;
