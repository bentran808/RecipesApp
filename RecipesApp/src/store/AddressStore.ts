import { toJS } from 'mobx';
import { types, destroy, SnapshotOut, getRoot, applySnapshot } from 'mobx-state-tree';
import { RootStore } from 'store/store';

const AddressEntry = types
  .model('AddressEntry', {
    id: types.optional(types.identifierNumber, 0),
    type: types.optional(types.string, ''),
    address: types.optional(types.string, '')
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).address.removeItem(self);
    },
    edit(type: string, address: string) {
      self.type = type;
      self.address = address;
    }
  }));

export type AddressModel = SnapshotOut<typeof AddressEntry>;

const AddressStore = types
  .model('', {
    items: types.array(AddressEntry),
    itemUsing: types.optional(AddressEntry, {})
  })
  .volatile(() => ({
    address: ''
  }))
  .actions((self) => ({
    setAddress(address: string) {
      self.address = address;
    },
    addNewItem(type: string, address: string) {
      const id = self.items.reduce((maxId, item) => Math.max(item.id, maxId), -1) + 1;
      self.items.push({ id, type, address });
    },
    removeItem(item: AddressModel) {
      destroy(item);
    },
    existAddress(address?: AddressModel) {
      return self.items.find((item) => item.id === address?.id);
    },
    setItemUsing(item: AddressModel) {
      applySnapshot(self, {
        ...self,
        itemUsing: item
      });
    }
  }))
  .views((self) => ({
    get addresses() {
      return toJS(self.items);
    }
  }));

export default AddressStore;
