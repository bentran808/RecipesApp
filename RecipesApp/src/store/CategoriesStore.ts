import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';
import { transformCategories } from 'utils';
import { RecipesEntry } from './RecipesStore';

export const CategoriesEntry = types.model('CategoriesEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string,
  recipes: types.maybe(types.array(RecipesEntry))
});

export type CategoryModel = SnapshotOut<typeof CategoriesEntry>;

const CategoriesStore = types
  .model('CategoriesStore', {
    lists: types.optional(types.array(CategoriesEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get categories() {
      return self.lists;
    },
    get categoriesJS() {
      return transformCategories(toJS(this.categories));
    }
  }))
  .actions((self) => ({
    fetchCategories: flow(function* fetchCategories() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchCategoriesRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        console.log(error);
        self.state = 'error';
      }
    })
  }));

export default CategoriesStore;
