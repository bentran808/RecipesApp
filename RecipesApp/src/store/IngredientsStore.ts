import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';

export const IngredientsEntry = types.model('IngredientsEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string
});

export type IngredientModel = SnapshotOut<typeof IngredientsEntry>;

const IngredientsStore = types
  .model('IngredientsStore', {
    lists: types.optional(types.array(IngredientsEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get ingredients() {
      return self.lists;
    },
    get ingredientsJS() {
      return toJS(this.ingredients);
    }
  }))
  .actions((self) => ({
    fetchIngredients: flow(function* fetchIngredients() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchIngredientsRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        self.state = 'error';
      }
    })
  }));

export default IngredientsStore;
