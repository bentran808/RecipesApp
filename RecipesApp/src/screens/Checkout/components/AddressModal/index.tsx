import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import AppStyles from 'AppStyles';
import Button from 'components/Button';
import Screens from 'constants/Screens';
import { useStore } from 'context';
import { observer } from 'mobx-react-lite';
import React from 'react';
import isEqual from 'react-fast-compare';
import { FlatList, Modal, Text, View } from 'react-native';
import { AddressModel } from 'store/AddressStore';
import styles from './styles';

type AddressModalNavigationProp = NativeStackNavigationProp<RootStackParamList>;
type Props = {
  navigation: AddressModalNavigationProp;
  modalVisible: boolean;
  onToggleModal: () => void;
  renderSeparator: () => JSX.Element;
  data: AddressModel[];
};

const AddressModal = ({
  navigation,
  data,
  modalVisible,
  onToggleModal,
  renderSeparator
}: Props) => {
  const { address } = useStore();

  const handlePressChooseAddress = (item: AddressModel) => () => {
    onToggleModal();
    address.setItemUsing(item);
  };

  const handlePressAdd = () => {
    navigation.navigate(Screens.Address.name as 'Address');
  };

  const renderHeader = () => <Text>Choose your address for delivery</Text>;

  const renderEmptyList = () => (
    <Button
      testID="addAddressBtn"
      type="contained"
      title="Add New Address"
      color="red"
      paddingHorizontal={30}
      onPress={handlePressAdd}
      style={{
        marginTop: 20
      }}
    />
  );

  const renderAddressItem = ({ item }: { item: AddressModel }) => (
    <View style={styles.addressWrapper}>
      <View>
        <Text style={AppStyles.textBold}>{item.type}</Text>
        <Text>{item.address}</Text>
      </View>
      <Button
        testID="chooseAddressBtn"
        type="outlined"
        color="red"
        borderColor="red"
        paddingVertical={5}
        paddingHorizontal={20}
        title="Choose"
        onPress={handlePressChooseAddress(item)}
      />
    </View>
  );
  const renderKeyExtractor = (item: AddressModel, index: number) => `${item.type}-${index}`;

  return (
    <Modal
      animationType="fade"
      transparent={true}
      visible={modalVisible}
      onRequestClose={onToggleModal}
    >
      <View style={styles.centeredView}>
        <View style={styles.modalView}>
          <FlatList
            showsVerticalScrollIndicator={false}
            data={data}
            ItemSeparatorComponent={renderSeparator}
            ListHeaderComponent={renderHeader}
            ListFooterComponent={renderEmptyList}
            renderItem={renderAddressItem}
            keyExtractor={renderKeyExtractor}
          />
        </View>
      </View>
    </Modal>
  );
};

export default React.memo(observer(AddressModal), isEqual);
