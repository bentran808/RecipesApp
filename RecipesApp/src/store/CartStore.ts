import { toJS } from 'mobx';
import { applySnapshot, destroy, getRoot, SnapshotOut, types } from 'mobx-state-tree';
import { RecipeModel, RecipesEntry } from 'store/RecipesStore';
import { RootStore } from 'store/store';

export const CartEntry = types
  .model('CartEntry', {
    item: RecipesEntry,
    quantity: types.optional(types.number, 1),
    inCart: types.optional(types.boolean, true)
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).cart.removeItem(self);
    },
    increase(quantity?: number) {
      self.quantity = quantity ? self.quantity + quantity : ++self.quantity;
    },
    decrease() {
      self.quantity === 1 ? this.remove() : --self.quantity;
    },
    toggle() {
      self.inCart = !self.inCart;
    }
  }));

const BillEntry = types.model('BillEntry', {
  type: types.optional(types.string, ''),
  name: types.optional(types.string, ''),
  code: types.optional(types.string, ''),
  price: types.optional(types.number, 0)
});

const OrderedEntry = types.model('OrderedEntry', {
  items: types.array(
    types.model({
      name: types.optional(types.string, ''),
      quantity: types.optional(types.number, 0)
    })
  ),
  total: types.number,
  createdAt: types.string
});

export type CartModel = SnapshotOut<typeof CartEntry>;
export type BillModel = SnapshotOut<typeof BillEntry>;
export type OrderedModel = SnapshotOut<typeof OrderedEntry>;

const CartStore = types
  .model('CartStore', {
    items: types.array(CartEntry),
    billItems: types.array(BillEntry),
    orderedItems: types.array(OrderedEntry)
  })
  .views((self) => ({
    get recipes() {
      return toJS(self.items);
    },
    get recipesInCart() {
      return this.recipes.filter((item) => item.inCart);
    },
    get inCartCount() {
      return this.recipesInCart.length;
    },
    get total() {
      return this.recipesInCart.reduce((sum, item) => sum + item.quantity * 20, 0);
    },
    get billDetails() {
      return toJS(self.billItems);
    },
    get toPay() {
      return this.billDetails.reduce(
        (sum, item) => (item.type === 'discount' ? sum - item.price : sum + item.price),
        0
      );
    },
    get orderedList() {
      return toJS(self.orderedItems);
    }
  }))
  .volatile(() => ({
    discountInput: '',
    invalidCoupon: false
  }))
  .actions((self) => ({
    addToCart(item: RecipeModel, quantity?: number) {
      const itemInCart = this.hasInCart(item);
      itemInCart ? itemInCart.increase(quantity) : self.items.push({ item, quantity });
    },
    removeItem(item: CartModel) {
      destroy(item);
    },
    hasInCart(recipe: RecipeModel) {
      return self.items.find((item) => item.item.id === recipe.id);
    },
    setDiscountInput(text: string) {
      if (!text) {
        self.invalidCoupon = false;
      }
      self.discountInput = text;
    },
    applyCoupon(item: BillModel) {
      const existItem = self.billItems.findIndex((bill) => bill.type === 'discount');
      if (existItem > -1) {
        self.billItems.splice(existItem, 1, item);
      } else {
        self.billItems.push(item);
      }
    },
    payment(order: OrderedModel) {
      self.orderedItems.unshift(order);
      applySnapshot(self, {
        ...self,
        items: [],
        billItems: []
      });
    },
    isInvalidCoupon(status: boolean) {
      self.invalidCoupon = status;
    }
  }));

export default CartStore;
