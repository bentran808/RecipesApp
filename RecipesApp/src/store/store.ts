import { Instance, types } from 'mobx-state-tree';
import RecipesStore from 'store/RecipesStore';
import CategoriesStore from './CategoriesStore';

const RootStore = types.model('RootStore', {
  categories: CategoriesStore,
  recipes: RecipesStore
});

const store = RootStore.create({
    categories: {},
    recipes: {}
});

export type RootStoreModel = Instance<typeof RootStore>;

export default store;
