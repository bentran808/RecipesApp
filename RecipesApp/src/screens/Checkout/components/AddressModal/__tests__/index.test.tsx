import { fireEvent, render } from '@testing-library/react-native';
import { StoreProvider } from 'context';
import { address } from 'mocks';
import React from 'react';
import renderer from 'react-test-renderer';
import store from 'store/store';
import AddressModal from '..';

describe('Address Modal Component', () => {
  const navigation: any = {
    navigate: jest.fn()
  };
  const props = {
    navigation,
    data: address,
    modalVisible: true,
    onToggleModal: jest.fn(),
    renderSeparator: jest.fn().mockReturnValue(null)
  };

  test('should render correctly', () => {
    const tree = renderer
      .create(
        <StoreProvider value={store}>
          <AddressModal {...props} />
        </StoreProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should call function handlePressAdd', () => {
    const { getByTestId } = render(
      <StoreProvider value={store}>
        <AddressModal {...props} />
      </StoreProvider>
    );
    const button = getByTestId('addAddressBtn');
    fireEvent.press(button);

    expect(navigation.navigate).toHaveBeenCalled();
  });

  test('should call function handlePressChooseAddress', () => {
    const { getAllByTestId } = render(
      <StoreProvider value={store}>
        <AddressModal {...props} />
      </StoreProvider>
    );
    const button = getAllByTestId('chooseAddressBtn')[0];
    fireEvent.press(button);

    expect(props.onToggleModal).toHaveBeenCalled();
  });
});
