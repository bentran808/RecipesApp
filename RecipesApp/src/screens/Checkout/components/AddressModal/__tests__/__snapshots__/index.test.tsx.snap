// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address Modal Component should render correctly 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,0.5)",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 20,
          "elevation": 5,
          "paddingVertical": 20,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 4,
          "width": 700,
        }
      }
    >
      <RCTScrollView
        ItemSeparatorComponent={
          [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "highlighted": false,
                  "leadingItem": Object {
                    "address": "448 TNV",
                    "id": 0,
                    "type": "Home",
                  },
                },
                Object {},
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": null,
              },
            ],
          }
        }
        ListFooterComponent={[Function]}
        ListHeaderComponent={[Function]}
        data={
          Array [
            Object {
              "address": "448 TNV",
              "id": 0,
              "type": "Home",
            },
            Object {
              "address": "604 NT",
              "id": 1,
              "type": "Work",
            },
          ]
        }
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
          >
            <Text>
              Choose your address for delivery
            </Text>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingVertical": 10,
                  "width": 650,
                }
              }
            >
              <View>
                <Text
                  style={
                    Object {
                      "fontWeight": "bold",
                    }
                  }
                >
                  Home
                </Text>
                <Text>
                  448 TNV
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="chooseAddressBtn"
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderRadius": 100,
                        "marginHorizontal": 10,
                      },
                      Object {
                        "borderColor": "red",
                        "borderWidth": 1,
                        "paddingHorizontal": 20,
                        "paddingVertical": 5,
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "red",
                          "fontSize": 14,
                          "fontWeight": "normal",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    Choose
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingVertical": 10,
                  "width": 650,
                }
              }
            >
              <View>
                <Text
                  style={
                    Object {
                      "fontWeight": "bold",
                    }
                  }
                >
                  Work
                </Text>
                <Text>
                  604 NT
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="chooseAddressBtn"
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderRadius": 100,
                        "marginHorizontal": 10,
                      },
                      Object {
                        "borderColor": "red",
                        "borderWidth": 1,
                        "paddingHorizontal": 20,
                        "paddingVertical": 5,
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "red",
                          "fontSize": 14,
                          "fontWeight": "normal",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    Choose
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="addAddressBtn"
            >
              <View
                style={
                  Array [
                    Object {
                      "borderRadius": 100,
                      "marginHorizontal": 10,
                    },
                    Object {
                      "backgroundColor": "red",
                      "paddingHorizontal": 30,
                      "paddingVertical": 10,
                    },
                    Object {
                      "marginTop": 20,
                    },
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 14,
                        "fontWeight": "normal",
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Add New Address
                </Text>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</Modal>
`;
